### PR TITLE
Add change address

### PR DESCRIPTION
**Pull request by Duddino**
## Abstract

Add proper change address, deriving them using `m/44'/119'/0'/1/${nAccount}` (or equivalent for testnet/hardware wallets).
All "internal" addresses, i.e. addresses that are not presented to the user, should derive this way.

## Testing
- Test send and receiving and that balance is correct
- (Advanced) Check that the explorer return the correct bip32 path

### CODE REVIEW

File: scripts/wallet.js [hunk 1]
```diff
@@ -215,25 +222,31 @@ export class Wallet {
...
+        for (const nReceiving of [0, 1]) {
+            const maxIndex = Math.max(
+                getNetwork().getLastWallet(nReceiving),
+                this.#addressIndeces.get(nReceiving) ?? 0
+            );
+
+            if (this.isHD()) {
+                for (let i = 0; i <= maxIndex + MAX_ACCOUNT_GAP; i++) {
+                    const path = this.getDerivationPath(nReceiving, i);
+                    console.log(nReceiving, path);
```
Comment (panleone) [inline scripts/wallet.js:236]: remove log

File: scripts/wallet.js [hunk 2]
```diff
@@ -46,9 +46,13 @@ export class Wallet {
...
     #nAccount;
     /**
-     * @type {number}
+     * Maps nReceiving->index of generated address
+     * nReceiving = 0 are regular addresses given to the user
+     * nReceiving = 1 are change addresses
+     * @type {Map<number, number>}
      */
-    #addressIndex = 0;
+    #addressIndeces = new Map();
```
Comment (panleone) [inline scripts/wallet.js:54]: typo: indexes

**Review by panleone — CHANGES_REQUESTED** [review]
a couple of nits

File: scripts/wallet.js [hunk 3]
```diff
@@ -215,25 +222,30 @@ export class Wallet {
...
-            for (let i = 0; i <= this.#addressIndex + MAX_ACCOUNT_GAP; i++) {
-                const path = this.getDerivationPath(0, i);
-                const testAddress = await this.#masterKey.getAddress(path);
-                if (address === testAddress) {
-                    this.#ownAddresses.set(address, path);
-                    return path;
+        // We only check 0, 1 because that's what blockbook does too by default
+        // https://github.com/trezor/blockbook/blob/master/docs/api.md#get-xpub
+        for (const nReceiving of [0, 1]) {
+            const maxIndex = Math.max(
```
Comment (panleone) [inline scripts/wallet.js:228]: update `addressIndices ` with the value given by the network (if `addressIndices.at(nReceiving)` is smaller)

**Review by panleone — CHANGES_REQUESTED** [review]

File: scripts/wallet.js (hunk 3 above)
Comment (Duddino) [reply scripts/wallet.js:228]: I feel like this shouldn't be in `isOwnAddress`, besides `getNewAddress` does this too

Comment (panleone) [reply scripts/wallet.js:228]: ok but then `addressIndices` will not be updated until you generate a new address

**Review by panleone — DISMISSED** [review]
tACK f2e06936c5f8b3681b20ff073196c95384306568
working fine on my hd wallet

**Comment by Duddino** [comment]
Closing in favor of a rewrite, as it's easier than solving conflicts

**Pull request closed without merging**